# TECtools_public

TECtools is a suite of tools for processing data from Transcription Elongation Complex RNA structure probing (TECprobe) data including:

**cotrans_preprocessor** - Perform sequencing read preprocessing to prepare data for analysis by ShapeMapper2

**mkmtrx** - Assemble data into matrix and other useful formats, reports alignment rates, and can be used to generate rdat files

**mtrx2cols** - Extract reactivity trajectories for specific nucleotides from a reactivity matrix

**TECtools is only intended for use on Linux and MacOS systems**

## Compiling TECtools scripts

**TECtools will only run on Linux and MacOS systems. Windows is not currently supported **

The cotrans_preprocessor, mkmtrx, and mtrx2cols scripts can be compiled using the following commands:

```
cd ./TECtools_public/build_TECprobe

sh ./build_TECprobe.sh
```

This will generate the executables cotrans_preprocessor, mkmtrx, and mtrx2cols in the build_TECprobe directory.

## Processing TECprobe data using cotrans_preprocessor

cotrans_preprocessor performs sequencing read preprocessing to prepare data for analysis by ShapeMapper 2

### Set cotrans_preprocessor run mode (required)

```
-m/--mode <run_mode_specifier>

run_mode_specifiers:

  MAKE_FASTA          Generate FASTA file to be used for target generation
  
  MAKE_3pEND_TARGETS  Generate 3' end targets to be used for demultiplexing FASTQ files by 
                      transript length, and intermediate transcript targets to be used for 
                      sequencing read alignment.
  
  PROCESS_MULTI       Perform preprocessing for TECprobe-ML experiments
  
  PROCESS_SINGLE      Perform preprocessing for TECprobe-SL experiments
  
  MAKE_RUN_SCRIPT      Generate shell script for running ShapeMapper2
```

### MAKE_FASTA mode inputs and options

Required:

```
-n/--seq-name <name_of_sequence>  Target name.

-s/--sequence <sequence>          Target sequence.
```

### MAKE_3pEnd_TARGETS mode inputs and options

Required:
```
-A/--fasta  <fasta_file>  Target sequence in FASTA format. Can be generated by running.
                          cotrans_preprocessor in MAKE_FASTA mode.
```

Optional:
```
-E/--end-length <n>     Length of 3' end targets. Default=14, 3' end target length should ideally
                        be 14-16 nt, because the RNA polymerase footprint on RNA is 14-15 nt.
                        
-S/--min-length <n>     Minimum intermediate transcript target length. Default=20.

-T/--test-data          Generate test data to be used for validating transcript 3' end mapping.
                        By default, generates 1000 test_data reads per intermediate transcript
                        using native sequence with randomized channel barcodes.
                        
-U/--multiplier <n>     Multiply the number of test_data reads generated by <n>.

-R/--end-randomization  Make random 1 nts substitutions, insertions, and deletions in the last
                        <end_length> nts of test_data reads. End-randomization should be used
                        to validate correct 3' end mapping. Complete coverage of all 1 nt
                        variations typically requires that --multiplier be set to at least 30.
```

### PROCESS_MULTI and PROCESS_SINGLE mode inputs and options

Required for PROCESS_MULTI and PROCESS_SINGLE modes:

```
-i/--read1 <read1_fastq>      Read 1 FASTQ file input.

-I/--read2 <read2_fastq>      Read 2 FASTQ file input.
```

Required for PROCESS_MULTI mode only:

```
-e/--3pEnd <3'_end_targets>   3' end targets file input.
```

Required for PROCESS_SINGLE mode only:

```
-a/--fasta-ref <fasta_file>   FASTA file containing single target.
```

Optional:

```
-p/--fastp-path <fastp_path>  Path to fastp executable

-t/--testdata                 Enable test data analysis. This option should only be used
                              when analyzing test_data sequencing reads that were 
                              generated in MAKE_3pEND_TARGETS mode using the -T option.
```

### MAKE_RUN_SCRIPT mode input

Required:
```
-c/--config <config_file>     Run script configuration file input.
```
