## Reactivity-constrained prediction of RNA folding intermediate structures using draw_intermediates

`draw_intermediates` performs reactivity-constrained minimum free energy RNA structure prediction for intermediate transcripts using the RNAstructure (Reuter, J. S. & Mathews, D. H., BMC Bioinformatics) Fold algorithm. `draw intermediates` requires the RNAstructure command line tool package, which is available at https://rna.urmc.rochester.edu/RNAstructure.html.

### draw_intermediates inputs and options
  
Required:
```
-r/--rna_structure <RNAstructure-path>  Path to RNAstructure command line tools directory
-m/--matrix <reactivity_matrix>         Input reactivity matrix csv file generated by the generate_cotrans_heatmap program
-s/--sequence <sequence_input>          Target RNA sequence in FASTA format
-p/--probe <probe_type>                 Chemical probe type. Accepted inputs are `SHAPE` and `DMS`
-o/--output <output_directory-name>     Output directory name.
```
 
Optional:
```
-a/--start-nt <n>   First nucleotide to include during structure prediction. Must be an integer.
```

### Basic usage of assemble_TECprobeLM_data

1. Run the command `export DATAPATH=<path_to_data_tables>`, where `<path_to_data_tables>` is the path to the `data_tables` directory within the RNAstructure command line tools directory.

2. Run the command:

`assemble_TECprobeLM_data -r <RNAstructure-path> -m <reactivity_matrix> -s <sequence_input> -p <probe_type> -o <output_directory>`

This will generate an output directory in which sub-directories for each transcript length contain:
  - The FASTA file that contains the sequence used for RNA structure prediction
  - The reactivity constraint file that was used for RNA structure prediction
  - The connectivity table that describes the predicted RNA structures
  - A postscript file that contains secondary structures of the predicted RNA structures.
  
